clippy_lints/src/unintentional_unit_return.rs

```diff
@@ -0,0 +1,155 @@
+    generics.predicates.iter().find_map(|(proj_pred, _)| {
+        if let Predicate::Projection(proj_pred) = proj_pred {
+            let projection_pred = cx.tcx.erase_late_bound_regions(proj_pred);
+            if projection_pred.projection_ty.substs == pred.trait_ref.substs {
+                return Some(projection_pred);
+            }
+        }
+        None
+    })
+}
+
+fn get_args_to_check<'tcx>(cx: &LateContext<'_, 'tcx>, expr: &'tcx Expr<'tcx>) -> Vec<(usize, String)> {
+    let mut args_to_check = Vec::new();
+    if let Some(def_id) = cx.tables.type_dependent_def_id(expr.hir_id) {
+        let fn_sig = cx.tcx.fn_sig(def_id);
+        let generics = cx.tcx.predicates_of(def_id);
+        let fn_mut_preds = get_predicates_for_trait_path(cx, generics, &paths::FN_MUT);
+        let ord_preds = get_predicates_for_trait_path(cx, generics, &paths::ORD);
+        let partial_ord_preds = get_predicates_for_trait_path(cx, generics, &paths::PARTIAL_ORD);
+        // Trying to call erase_late_bound_regions on fn_sig.inputs() gives the following error
+        // The trait `rustc::ty::TypeFoldable<'_>` is not implemented for `&[&rustc::ty::TyS<'_>]`
+        let inputs_output = cx.tcx.erase_late_bound_regions(&fn_sig.inputs_and_output());
+        inputs_output.iter().enumerate().for_each(|(i, inp)| {
+            // Ignore output param
+            if i == inputs_output.len() - 1 {
+                return;
+            }
+            fn_mut_preds.iter().for_each(|pred| {
+                if_chain! {
+                    if let Some(trait_pred) = unwrap_trait_pred(cx, pred);
+                    if trait_pred.self_ty() == *inp;
+                    if let Some(return_ty_pred) = get_projection_pred(cx, generics, trait_pred);
+                    then {
+                        if ord_preds.iter().any(|ord|  {
+                            unwrap_trait_pred(cx, ord).unwrap().self_ty()  == return_ty_pred.ty
+                        }) {
+                            args_to_check.push((i, "Ord".to_string()));
+                        }
+                        else if partial_ord_preds.iter().any(|pord| {
+                            unwrap_trait_pred(cx, pord).unwrap().self_ty()  == return_ty_pred.ty
+                        }) {
+                            args_to_check.push((i, "PartialOrd".to_string()));
+                        }
+                    }
+                }
+            });
+        });
+    }
+    args_to_check
+}
+
+fn check_arg<'tcx>(cx: &LateContext<'_, 'tcx>, arg: &'tcx Expr<'tcx>) -> Option<&'tcx Stmt<'tcx>> {
+    if let ExprKind::Closure(_, _fn_decl, body_id, _span, _) = arg.kind {
+        if_chain! {
+            let body = cx.tcx.hir().body(body_id);
+            if let ExprKind::Block(block, _) = body.value.kind;
+            if let Some(stmt) = block.stmts.last();
+            if let StmtKind::Semi(_) = stmt.kind;
+            then { return Some(stmt) }
+        }
+    }
+    return None;
```

Comment:
Can be simplified to 
```suggestion
    if_chain! {
        if let ExprKind::Closure(_, _fn_decl, body_id, _span, _) = arg.kind;
        let body = cx.tcx.hir().body(body_id);
        if let ExprKind::Block(block, _) = body.value.kind;
        if let Some(stmt) = block.stmts.last();
        if let StmtKind::Semi(_) = stmt.kind;
        then { 
            Some(stmt) 
        } else {
            None
        }
    }
```